pkg/resourcemonitor/noderesourcesaggregator.go

```diff
@@ -0,0 +1,274 @@
+	perNuma := make(map[int]map[v1.ResourceName]*resourceData)
+	for nodeID, nodeRes := range noderesourceData.perNUMACapacity {
+		perNuma[nodeID] = make(map[v1.ResourceName]*resourceData)
+		for resName, resCap := range nodeRes {
+			perNuma[nodeID][resName] = &resourceData{
+				capacity:    resCap,
+				allocatable: resCap,
+			}
+		}
+	}
+
+	for _, podRes := range podResData {
+		for _, contRes := range podRes.Containers {
+			for _, res := range contRes.Resources {
+				noderesourceData.updateAllocatable(perNuma, res)
+			}
+		}
+	}
+
+	// zones := make([]topologyv1alpha1.Zone, 0)
+	zones := make(topologyv1alpha1.ZoneList, 0)
+	for nodeID, resList := range perNuma {
+		zone := topologyv1alpha1.Zone{
+			Name:      makeZoneName(nodeID),
+			Type:      "Node",
+			Resources: make(topologyv1alpha1.ResourceInfoList, 0),
+		}
+
+		costs, err := makeCostsPerNumaNode(noderesourceData.topo.Nodes, nodeID)
+		if err != nil {
+			log.Printf("cannot find costs for NUMA node %d: %v", nodeID, err)
+		} else {
+			zone.Costs = topologyv1alpha1.CostList(costs)
+		}
+
+		for name, resData := range resList {
+			allocatableQty := *resource.NewQuantity(resData.allocatable, resource.DecimalSI)
+			capacityQty := *resource.NewQuantity(resData.capacity, resource.DecimalSI)
+			zone.Resources = append(zone.Resources, topologyv1alpha1.ResourceInfo{
+				Name:        name.String(),
+				Allocatable: intstr.FromString(allocatableQty.String()),
+				Capacity:    intstr.FromString(capacityQty.String()),
+			})
+		}
+		zones = append(zones, zone)
+	}
+	return zones
+}
+
+// GetContainerDevicesFromAllocatableResources normalize all compute resources to ContainerDevices.
+// This is helpful because cpuIDs are not represented as ContainerDevices, but with a different format;
+// Having a consistent representation of all the resources as ContainerDevices makes it simpler for
+func GetContainerDevicesFromAllocatableResources(availRes *podresourcesapi.AllocatableResourcesResponse, topo *ghw.TopologyInfo) []*podresourcesapi.ContainerDevices {
+	var contDevs []*podresourcesapi.ContainerDevices
+	for _, dev := range availRes.GetDevices() {
+		contDevs = append(contDevs, dev)
+	}
+
+	cpuIDToNodeIDMap := MakeLogicalCoreIDToNodeIDMap(topo)
+
+	cpusPerNuma := make(map[int][]string)
+	for _, cpuID := range availRes.GetCpuIds() {
+		nodeID, ok := cpuIDToNodeIDMap[int(cpuID)]
+		if !ok {
+			log.Printf("cannot find the NUMA node for CPU %d", cpuID)
+			continue
+		}
+
+		cpuIDList := cpusPerNuma[nodeID]
+		cpuIDList = append(cpuIDList, fmt.Sprintf("%d", cpuID))
+		cpusPerNuma[nodeID] = cpuIDList
+	}
+
+	for nodeID, cpuList := range cpusPerNuma {
+		contDevs = append(contDevs, &podresourcesapi.ContainerDevices{
+			ResourceName: string(v1.ResourceCPU),
+			DeviceIds:    cpuList,
+			Topology: &podresourcesapi.TopologyInfo{
+				Nodes: []*podresourcesapi.NUMANode{
+					{ID: int64(nodeID)},
+				},
+			},
+		})
+	}
```

Comment:
Hmm, probably I'm just dumb, but somehow all the aggregator code is surprisingly hard to read/comprehend for such a simple accounting scenario (this currently handles only numa nodes, no other "topology levels" like cores, dies or sockets). Maybe re-thinking the data structures could help - e.g. not using internally the somewhat awkward representation of resources from podresource api(?)

Reply:
Please refer to my comment here: https://github.com/kubernetes-sigs/node-feature-discovery/pull/525#discussion_r698841568 and let me know what you think.